src/nvim/eval.c

```diff
@@ -13358,24 +13358,56 @@ static void get_maparg(typval_T *argvars, typval_T *rettv, int exact)
     rettv_dict_alloc(rettv);
     if (rhs != NULL) {
       // Return a dictionary.
-      char_u *lhs = str2special_save(mp->m_keys, true);
-      char *const mapmode = map_mode_to_chars(mp->m_mode);
-      dict_T *dict = rettv->vval.v_dict;
+      mapblock_fill_dict(rettv->vval.v_dict, mp, buffer_local, true);
+    }
+  }
+}
 
-      dict_add_nr_str(dict, "lhs",     0L, lhs);
-      dict_add_nr_str(dict, "rhs",     0L, mp->m_orig_str);
-      dict_add_nr_str(dict, "noremap", mp->m_noremap ? 1L : 0L, NULL);
-      dict_add_nr_str(dict, "expr",    mp->m_expr    ? 1L : 0L, NULL);
-      dict_add_nr_str(dict, "silent",  mp->m_silent  ? 1L : 0L, NULL);
-      dict_add_nr_str(dict, "sid",     (long)mp->m_script_ID, NULL);
-      dict_add_nr_str(dict, "buffer",  (long)buffer_local, NULL);
-      dict_add_nr_str(dict, "nowait",  mp->m_nowait  ? 1L : 0L, NULL);
-      dict_add_nr_str(dict, "mode",    0L, (char_u *)mapmode);
+// @param dict The dictionary to be filled
+// @param mp The maphash that contains the mapping information
+// @param buffer_id
+//      -1: Global mapping
+//      0: Current buffer
+//      else: The desired buffer ID
+//          Assumed to be a valid buffer ID
+//  @param compatible True for compatible with old maparg() dict
+void mapblock_fill_dict(dict_T *const dict, const mapblock_T *const mp, long buffer_id, bool compatible)
```

Comment:
why not simply send in the whatever value `"buffer"` should be set to?  buffer_id is not used otherwise inside this function. `get_maparg` can just continue to send in `buffer_local` and if the `nvim_buf` function does `find_buffer_by_handle` directly like the others, the `buffer==0` logic can stay there and need not be replicated.